### PR TITLE
HttpInput read to ByteBuffer

### LINKED ISSUE #7688: Read data to native memory from HttpInput
**Target Jetty version(s)**
jetty-11.0.9
**Enhancement Description**
Our purpose is read data direct to native memory from jetty direct buffers, then we could do CPU heavy compute with native lib without copy data between java heap and native heap.

Jetty has provide flexibility let us inherit class and implement our special code. The problem is some fields/methods in HttpChannel and HttpInput are private and can't be accessed in sub class.

The change is explore necessary methods to sub class. I will prepare a PR to show the change.

### PR DESCRIPTION
**Pull request by jiang-cao**
add int read(ByteBuffer buffer) in HttpInput.

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java [hunk 1]
```diff
@@ -248,6 +269,11 @@ public int read() throws IOException
 
     @Override
     public int read(byte[] b, int off, int len) throws IOException
+    {
+        return read(ByteBuffer.wrap(b, off, len));
```
Comment (jiang-cao) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java:273]: The main flow in read(byte[] b, int off, int len)  is same as read(ByteBuffer buffer).  Wrap the array can avoid these duplicate code.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java [hunk 2]
```diff
@@ -130,9 +130,30 @@ public void addInterceptor(Interceptor interceptor)
         }
     }
 
-    private int get(Content content, byte[] bytes, int offset, int length)
+    private int get(Content content, ByteBuffer des)
     {
-        int consumed = content.get(bytes, offset, length);
+        int consumed;
+        if (des.hasArray())
```
Comment (jiang-cao) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java:136]: May I delete the des.hasArray branch?  The else branch can support same functionality.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java [hunk 3]
```diff
@@ -239,7 +260,7 @@ public int read() throws IOException
     {
         try (AutoLock lock = _contentProducer.lock())
         {
-            int read = read(_oneByteBuffer, 0, 1);
+            int read = read(ByteBuffer.wrap(_oneByteBuffer, 0, 1));
```
Comment (jiang-cao) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java:263]: _oneByteBuffer can change to ByBuffer to avoid this wrap().

**Comment by sbordet** [comment]
@jiang-cao please base your work off the `jetty-10.0.x` branch, we will take care of merging to 11.

Please do not wrap `byte[]` into `ByteBuffer` because these methods are performance sensitive and we do not want to allocate a `ByteBuffer` for every read.

Let's start with adding the method you need, and if there are duplication happy to refactor, but let's not start by refactoring :)

Thanks!

**Comment by jiang-cao** [comment]
Thanks @sbordet 
I create https://github.com/eclipse/jetty.project/pull/7778 base on jetty-10.0.x.

**Pull request closed without merging**